api/src/com/cloud/agent/api/storage/OVFHelper.java

```diff
@@ -0,0 +1,333 @@
+                        units = 1024;
+                    } else if (allocationUnits.equalsIgnoreCase("MB") || allocationUnits.equalsIgnoreCase("MegaBytes") || allocationUnits.equalsIgnoreCase("byte * 2^20")) {
+                        units = 1024 * 1024;
+                    } else if (allocationUnits.equalsIgnoreCase("GB") || allocationUnits.equalsIgnoreCase("GigaBytes") || allocationUnits.equalsIgnoreCase("byte * 2^30")) {
+                        units = 1024 * 1024 * 1024;
+                    }
+                    od._capacity = od._capacity * units;
+                    s_logger.info("MDOVA getOVFVolumeInfo Disk _capacity  = " + od._capacity);
+                }
+                od._controller = getControllerType(items, od._diskId);
+                vd.add(od);
+            }
+
+        } catch (SAXException | IOException | ParserConfigurationException e) {
+            s_logger.error("Unexpected exception caught while parsing ovf file:" + ovfFilePath, e);
+            throw new CloudRuntimeException(e);
+        }
+
+        List<DatadiskTO> disksTO = new ArrayList<DatadiskTO>();
+        for (OVFFile of : vf) {
+            if (of._id == null || of._id.isEmpty()){
+                s_logger.error("The ovf file info is incomplete file info");
+                throw new CloudRuntimeException("The ovf file info has incomplete file info");
+            }
+            OVFDisk cdisk = getDisk(of._id, vd);
+            if (cdisk == null && !of._iso){
+                s_logger.error("The ovf file info has incomplete disk info");
+                throw new CloudRuntimeException("The ovf file info has incomplete disk info");
+            }
+            Long capacity = cdisk == null ? of._size : cdisk._capacity;
+            String controller = cdisk == null ? "" : cdisk._controller._name;
+            String controllerSubType = cdisk == null ? "" : cdisk._controller._subType;
+            String dataDiskPath = ovfFile.getParent() + File.separator + of._href;
+            s_logger.info("MDOVA getOVFVolumeInfo diskName = " + of._href + ", dataDiskPath = " + dataDiskPath);
+            File f = new File(dataDiskPath);
+            if (!f.exists() || f.isDirectory()) {
+                s_logger.error("One of the attached disk or iso does not exists " + dataDiskPath);
+                throw new CloudRuntimeException("One of the attached disk or iso as stated on OVF does not exists " + dataDiskPath);
+            }
+            disksTO.add(new DatadiskTO(dataDiskPath, capacity, of._size, of._id, of._iso, of._bootable, controller, controllerSubType));
+        }
+        //check if first disk is an iso move it to the end
+        DatadiskTO fd = disksTO.get(0);
+        if (fd.isIso()) {
+            disksTO.remove(0);
+            disksTO.add(fd);
+        }
+        return disksTO;
+    }
+
+    private OVFDiskController getControllerType(final NodeList itemList, final String diskId) {
```

Comment:
I am seeing these manual parses of XML and thinking, why not use a tool for that? 
It feels that it is being added code that is not needed.

Reply:
I agree, I'll refactor this soon

Reply:
@rafaelweingartner I think we expored but did not find any opensource ovf reading/writing/editing library. This implements a subset of a general ovf manupulating logic to support multidisk+iso ova. I think we can refactor this as next steps and explore options in future.